File: VDFEditor/src/main/java/de/treichels/hott/vdfeditor/ui/MessageDialog.java
package de.treichels.hott.vdfeditor.ui;

import java.awt.Desktop;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;

import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.w3c.dom.events.EventTarget;

import javafx.application.Platform;
import javafx.concurrent.Worker;
import javafx.scene.control.Alert;
import javafx.scene.web.WebEngine;
import javafx.scene.web.WebView;
import javafx.stage.Stage;

class MessageDialog extends Alert {
    static void show(final AlertType alertType, final String header, final boolean html, final String message, final Object... args) {
        Platform.runLater(() -> new MessageDialog(alertType, header, html, message, args).showAndWait());
    }

    static void show(final AlertType alertType, final String header, final String message, final Object... args) {
        Platform.runLater(() -> new MessageDialog(alertType, header, false, message, args).showAndWait());
    }

    MessageDialog(final AlertType alertType, final String header, final boolean html, final String message, final Object... args) {
        super(alertType);

        ((Stage) getDialogPane().getScene().getWindow()).getIcons().add(Controller.ICON);
        setHeaderText(header);

        final String text = String.format(message, args);

        if (html) {
            final WebView webView = new WebView();
            final WebEngine engine = webView.getEngine();
            engine.loadContent(text);
            webView.setPrefSize(400, 400);
            getDialogPane().setContent(webView);
            engine.getLoadWorker().stateProperty().addListener((observable, oldState, newState) -> {
                if (newState == Worker.State.SUCCEEDED) {
                    // loop though all <a> tags
                    final NodeList links = engine.getDocument().getElementsByTagName("a");
                    for (int i = 0; i < links.getLength(); i++) {
                        final Element link = (Element) links.item(i);
                        final String url = link.getAttribute("href");

                        // disable normal onclick event handler as it will load the page into the dialog
                        link.setAttribute("onclick", "return false;");

                        // instead, load the url using the system default web browser
                        ((EventTarget) link).addEventListener("click", ev -> browse(url), false);
                    }
                }
            });
        } else
            setContentText(text);
    }

    MessageDialog(final AlertType alertType, final String header, final String message, final Object... args) {
        this(alertType, header, false, message, args);
    }

    private void browse(final String url) {
        try {
            Desktop.getDesktop().browse(new URI(url));
        } catch (IOException | URISyntaxException e) {
            ExceptionDialog.show(e);
        }
    }
}

File: HoTT-Model/src/main/kotlin/de/treichels/hott/util/MessageDialog.kt
package de.treichels.hott.util

import javafx.application.Platform
import javafx.concurrent.Worker
import javafx.scene.control.Alert
import javafx.stage.Stage
import org.w3c.dom.Element
import org.w3c.dom.events.EventTarget
import tornadofx.*
import java.awt.Desktop
import java.net.URI

class MessageDialog(alertType: Alert.AlertType, header: String, html: Boolean, message: String, vararg args: Any) : Alert(alertType) {
    init {
        (dialogPane.scene.window as Stage).icons += ResourceLookup(MessageDialog).image("icon.png")
        headerText = header

        val text = String.format(message, *args)

        if (html) {
            dialogPane.content = webview {
                prefHeight = 400.0
                prefWidth = 400.0
                engine.loadContent(text)
                engine.loadWorker.stateProperty().addListener { _, _, newState ->
                    if (newState == Worker.State.SUCCEEDED) {
                        // loop though all <a> tags
                        val links = engine.document.getElementsByTagName("a")
                        for (i in 0 until links.length) {
                            val link = links.item(i) as Element
                            val url = link.getAttribute("href")

                            // disable normal onclick event handler as it will load the page into the dialog
                            link.setAttribute("onclick", "return false;")

                            // instead, load the url using the system default web browser
                            (link as EventTarget).addEventListener("click", { Desktop.getDesktop().browse(URI(url)) }, false)
                        }
                    }
                }
            }
        } else
            contentText = text
    }

    constructor(alertType: Alert.AlertType, header: String, message: String, vararg args: Any) : this(alertType, header, false, message, *args)

    companion object {
        fun show(alertType: Alert.AlertType, header: String, html: Boolean, message: String, vararg args: Any) {
            Platform.runLater { MessageDialog(alertType, header, html, message, *args).showAndWait() }
        }

        fun show(alertType: Alert.AlertType, header: String, message: String, vararg args: Any) {
            Platform.runLater { MessageDialog(alertType, header, false, message, *args).showAndWait() }
        }
    }
}
